File: greatwestern/src/notification.rs
// SPDX-FileCopyrightText: © 2024 Christopher Woods <Christopher.Woods@bristol.ac.uk>
// SPDX-License-Identifier: MIT

use crate::grammar::{ProjectIdentifier, UserIdentifier};
use crate::Hpc;

use serde::{Deserialize, Serialize};
use std::fmt;
use templemeads::notification::Notification;
use templemeads::Error;

/// A fire-and-forget event notification. Unlike a Job, a Notification is not stored
/// on any board, carries no state machine, and no result is returned to the sender.
/// Analogous to UDP vs TCP for Jobs.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum NotificationEvent {
    /// A user was successfully added to a system
    UserAdded(UserIdentifier),
    /// A user was removed from a system
    UserRemoved(UserIdentifier),
    /// A user's details were changed (e.g. home directory updated)
    UserChanged(UserIdentifier),
    /// A user was blocked from logging in
    UserBlocked(UserIdentifier),
    /// A previously blocked user was unblocked
    UserUnblocked(UserIdentifier),
    /// A project was added to a system
    ProjectAdded(ProjectIdentifier),
    /// A project was removed from a system
    ProjectRemoved(ProjectIdentifier),
    /// A project's details were changed
    ProjectChanged(ProjectIdentifier),
    /// All users in a project were blocked
    ProjectBlocked(ProjectIdentifier),
    /// All users in a project were unblocked
    ProjectUnblocked(ProjectIdentifier),
    /// An award (project) was created or registered in the web portal
    AwardAdded(ProjectIdentifier),
    /// An award (project) was removed from the web portal
    AwardRemoved(ProjectIdentifier),
    /// An award (project) was updated in the web portal
    AwardChanged(ProjectIdentifier),
    /// An award was accepted by the receiving portal
    AwardAccepted(ProjectIdentifier),
    /// An award was rejected by the receiving portal
    AwardRejected(ProjectIdentifier),
    /// Infrastructure-only: used by the bridge agent to ask the portal to forward
    /// an inner notification southbound, stripping the bridge from the path.
    /// Analogous to `Instruction::Submit` for Jobs. Not accepted by `parse()`.
    Forward(Box<Notification<Hpc>>),
}

impl NotificationEvent {
    pub fn parse(s: &str) -> Result<Self, Error> {
        let (event_name, rest) = match s.split_once(' ') {
            Some((e, r)) => (e, r.trim()),
            None => (s.trim(), ""),
        };

        match event_name {
            "user_added" => Ok(Self::UserAdded(UserIdentifier::parse(rest)?)),
            "user_removed" => Ok(Self::UserRemoved(UserIdentifier::parse(rest)?)),
            "user_changed" => Ok(Self::UserChanged(UserIdentifier::parse(rest)?)),
            "user_blocked" => Ok(Self::UserBlocked(UserIdentifier::parse(rest)?)),
            "user_unblocked" => Ok(Self::UserUnblocked(UserIdentifier::parse(rest)?)),
            "project_added" => Ok(Self::ProjectAdded(ProjectIdentifier::parse(rest)?)),
            "project_removed" => Ok(Self::ProjectRemoved(ProjectIdentifier::parse(rest)?)),
            "project_changed" => Ok(Self::ProjectChanged(ProjectIdentifier::parse(rest)?)),
            "project_blocked" => Ok(Self::ProjectBlocked(ProjectIdentifier::parse(rest)?)),
            "project_unblocked" => Ok(Self::ProjectUnblocked(ProjectIdentifier::parse(rest)?)),
            "award_added" => Ok(Self::AwardAdded(ProjectIdentifier::parse(rest)?)),
            "award_removed" => Ok(Self::AwardRemoved(ProjectIdentifier::parse(rest)?)),
            "award_changed" => Ok(Self::AwardChanged(ProjectIdentifier::parse(rest)?)),
            "award_accepted" => Ok(Self::AwardAccepted(ProjectIdentifier::parse(rest)?)),
            "award_rejected" => Ok(Self::AwardRejected(ProjectIdentifier::parse(rest)?)),
            "forward" => Err(Error::Parse(
                "NotificationEvent::Forward is an infrastructure-only event and cannot be parsed from a string".to_owned(),
            )),
            unknown => Err(Error::Parse(format!(
                "Unknown notification event: '{}'",
                unknown
            ))),
        }
    }
}

impl fmt::Display for NotificationEvent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::UserAdded(u) => write!(f, "user_added {}", u),
            Self::UserRemoved(u) => write!(f, "user_removed {}", u),
            Self::UserChanged(u) => write!(f, "user_changed {}", u),
            Self::UserBlocked(u) => write!(f, "user_blocked {}", u),
            Self::UserUnblocked(u) => write!(f, "user_unblocked {}", u),
            Self::ProjectAdded(p) => write!(f, "project_added {}", p),
            Self::ProjectRemoved(p) => write!(f, "project_removed {}", p),
            Self::ProjectChanged(p) => write!(f, "project_changed {}", p),
            Self::ProjectBlocked(p) => write!(f, "project_blocked {}", p),
            Self::ProjectUnblocked(p) => write!(f, "project_unblocked {}", p),
            Self::AwardAdded(p) => write!(f, "award_added {}", p),
            Self::AwardRemoved(p) => write!(f, "award_removed {}", p),
            Self::AwardChanged(p) => write!(f, "award_changed {}", p),
            Self::AwardAccepted(p) => write!(f, "award_accepted {}", p),
            Self::AwardRejected(p) => write!(f, "award_rejected {}", p),
            Self::Forward(n) => write!(f, "forward [{}]", n),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_notification_event_parse_display_roundtrip() {
        #[allow(clippy::unwrap_used)]
        let user = UserIdentifier::parse("chris.project.brics").unwrap();
        let event = NotificationEvent::UserAdded(user);
        let s = event.to_string();
        #[allow(clippy::unwrap_used)]
        let parsed = NotificationEvent::parse(&s).unwrap();
        assert_eq!(event, parsed);
    }

    #[test]
    fn test_all_notification_events() {
        #[allow(clippy::unwrap_used)]
        let cases = vec![
            "user_added chris.project.brics",
            "user_removed chris.project.brics",
            "user_changed chris.project.brics",
            "user_blocked chris.project.brics",
            "user_unblocked chris.project.brics",
        ];
        for case in cases {
            #[allow(clippy::unwrap_used)]
            let event = NotificationEvent::parse(case).unwrap();
            assert_eq!(event.to_string(), case);
        }
    }

    #[test]
    fn test_project_notification_events() {
        #[allow(clippy::unwrap_used)]
        let cases = vec![
            "project_added myproject.brics",
            "project_removed myproject.brics",
            "project_changed myproject.brics",
            "project_blocked myproject.brics",
            "project_unblocked myproject.brics",
        ];
        for case in cases {
            #[allow(clippy::unwrap_used)]
            let event = NotificationEvent::parse(case).unwrap();
            assert_eq!(event.to_string(), case);
        }
    }

    #[test]
    fn test_unknown_event_errors() {
        let result = NotificationEvent::parse("nonexistent_event foo.bar.brics");
        assert!(result.is_err());
    }
}

File: java/src/main/java/org/openportal/Notification.java
// SPDX-FileCopyrightText: © 2026 Christopher Woods <Christopher.Woods@bristol.ac.uk>
// SPDX-License-Identifier: MIT

package org.openportal;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.UUID;

/**
 * A fire-and-forget event from the network.
 *
 * <p>UDP to a {@link Job}'s TCP: no board, no state, no result, and no
 * acknowledgement. Something happened - a user was added, an award was accepted -
 * and you are being told. Nothing is retried on your behalf and nothing notices
 * if you ignore it.
 *
 * <p>Arrives one of two ways. The bridge signals your {@code notification_url}
 * with {@code ?notification_id=<uuid>} and you fetch it
 * ({@link BridgeClient#fetchNotification}), or the whole JSON is posted to you.
 * Either way {@link #eventType} is what to dispatch on and
 * {@link #eventArgument} carries the identifier it is about.
 *
 * <p>The event is an externally-tagged enum on the wire -
 * {@code {"UserAdded": "chris.p.portal"}} - which is why the string form
 * ({@code "user_added chris.p.portal"}) and the JSON form spell it differently.
 * Both are handled here; compare against {@link #eventType}, which is always
 * the snake_case name.
 */
public final class Notification {

    /** The domain whose event vocabulary this client speaks. */
    public static final String DOMAIN = "greatwestern";

    /** The domain release this client was written against. */
    public static final String DOMAIN_VERSION = "0.92.0";

    private final ObjectNode node;

    private Notification(ObjectNode node) {
        this.node = node;
    }

    /**
     * Parse the {@code "<destination> <event> [<argument>]"} string form.
     *
     * <p>Gets a fresh id, since the string form carries none.
     */
    public static Notification parse(String command) {
        if (command == null || command.isBlank()) {
            throw new IllegalArgumentException("a notification needs a destination and an event");
        }

        String[] parts = command.trim().split("\\s+", 3);

        if (parts.length < 2) {
            throw new IllegalArgumentException(
                    "Invalid notification - needs a destination and an event: '" + command + "'");
        }

        String eventType = parts[1];
        String argument = parts.length > 2 ? parts[2] : "";

        ObjectNode node = Json.object();
        node.put("id", UUID.randomUUID().toString());
        node.put("destination", parts[0]);
        node.putObject("event").put(variantFor(eventType), argument);

        // Stamped once, at construction, exactly as the Rust side does - the
        // event vocabulary belongs to a domain, and a peer speaking a different
        // one has no business reading it. Both fields are optional on the wire,
        // so a notification without them is still valid; this is what an
        // agent's own would carry.
        node.put("domain", DOMAIN);
        node.put("domain_version", DOMAIN_VERSION);

        return new Notification(node);
    }

    public static Notification fromJson(String json) {
        return fromJson(Json.parse(json));
    }

    public static Notification fromJson(JsonNode json) {
        if (json == null || !json.isObject()) {
            throw new IllegalArgumentException("a notification is a JSON object");
        }

        return new Notification((ObjectNode) json.deepCopy());
    }

    /** This notification's UUID. For logging - it is not stored anywhere. */
    public String id() {
        return node.path("id").asText();
    }

    /** Where the notification came from, as a dotted path. */
    public String destination() {
        return node.path("destination").asText();
    }

    /** The route, parsed. */
    public Destination destinationPath() {
        return Destination.parse(destination());
    }

    /** The event keyword alone, snake_case - {@code "user_added"}. Dispatch on this. */
    public String eventType() {
        JsonNode event = node.path("event");

        if (event.isTextual()) {
            // The string form, if a peer ever sends one: "user_added <arg>".
            String[] parts = event.asText().trim().split("\\s+", 2);

            return parts[0];
        }

        java.util.Iterator<String> names = event.fieldNames();

        return names.hasNext() ? snakeCase(names.next()) : "";
    }

    /** Everything after the keyword - usually an identifier. Empty if there is none. */
    public String eventArgument() {
        JsonNode event = node.path("event");

        if (event.isTextual()) {
            String[] parts = event.asText().trim().split("\\s+", 2);

            return parts.length > 1 ? parts[1] : "";
        }

        java.util.Iterator<JsonNode> values = event.elements();

        if (!values.hasNext()) {
            return "";
        }

        JsonNode argument = values.next();

        // A `Forward` carries a whole nested notification rather than an
        // identifier, so it has no text form - hand back its JSON.
        return argument.isTextual() ? argument.asText() : Json.write(argument);
    }

    /** The whole event, keyword and argument - {@code "user_added chris.p.portal"}. */
    public String event() {
        String argument = eventArgument();

        return argument.isEmpty() ? eventType() : eventType() + " " + argument;
    }

    /** The argument as a user identifier, for a {@code user_*} event. */
    public UserIdentifier user() {
        return UserIdentifier.parse(eventArgument());
    }

    /** The argument as a project identifier, for a {@code project_*} or {@code award_*} event. */
    public ProjectIdentifier projectOrAward() {
        return ProjectIdentifier.parse(eventArgument());
    }

    /** The domain that authored this event, if it said. */
    public java.util.Optional<String> domain() {
        return node.hasNonNull("domain")
                ? java.util.Optional.of(node.get("domain").asText())
                : java.util.Optional.empty();
    }

    /** That domain's version, if it said. */
    public java.util.Optional<String> domainVersion() {
        return node.hasNonNull("domain_version")
                ? java.util.Optional.of(node.get("domain_version").asText())
                : java.util.Optional.empty();
    }

    public JsonNode toJson() {
        return node.deepCopy();
    }

    /** The string form: {@code "<destination> <event>"}. */
    @Override
    public String toString() {
        return destination() + " " + event();
    }

    @Override
    public boolean equals(Object other) {
        return other instanceof Notification && node.equals(((Notification) other).node);
    }

    @Override
    public int hashCode() {
        return node.hashCode();
    }

    /**
     * The CamelCase variant name a snake_case event goes on the wire as.
     *
     * <p>Only the fifteen the domain defines are accepted; {@code forward} is
     * infrastructure-only and cannot be built from a string. Rejecting an
     * unknown name here rather than CamelCasing it blindly means a typo fails
     * where it is written instead of being delivered as an event nobody
     * handles.
     */
    private static String variantFor(String eventType) {
        return switch (eventType) {
            case "user_added" -> "UserAdded";
            case "user_removed" -> "UserRemoved";
            case "user_changed" -> "UserChanged";
            case "user_blocked" -> "UserBlocked";
            case "user_unblocked" -> "UserUnblocked";
            case "project_added" -> "ProjectAdded";
            case "project_removed" -> "ProjectRemoved";
            case "project_changed" -> "ProjectChanged";
            case "project_blocked" -> "ProjectBlocked";
            case "project_unblocked" -> "ProjectUnblocked";
            case "award_added" -> "AwardAdded";
            case "award_removed" -> "AwardRemoved";
            case "award_changed" -> "AwardChanged";
            case "award_accepted" -> "AwardAccepted";
            case "award_rejected" -> "AwardRejected";
            case "forward" -> throw new IllegalArgumentException(
                    "forward is an infrastructure-only event and cannot be parsed from a string");
            default -> throw new IllegalArgumentException(
                    "Unknown notification event: '" + eventType + "'");
        };
    }

    /** {@code "UserAdded"} to {@code "user_added"}, for whatever arrives. */
    private static String snakeCase(String variant) {
        StringBuilder text = new StringBuilder();

        for (int i = 0; i < variant.length(); i++) {
            char c = variant.charAt(i);

            if (Character.isUpperCase(c)) {
                if (i > 0) {
                    text.append('_');
                }

                text.append(Character.toLowerCase(c));
            } else {
                text.append(c);
            }
        }

        return text.toString();
    }
}
